File: perl/2015/day06b/day06b.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub init_grid() {
    my @grid;

    for my $row (0..999) {
        for my $col (0..999) {
            $grid[$row][$col] = 0;
        }
    }
    return \@grid;
}

sub perform {
    my ($grid, $action, $r1, $c1, $r2, $c2) = @_;

    for my $row ($r1..$r2) {
        for my $col ($c1..$c2) {
            if ( $action eq 'turn on' ) {
                $grid->[$row][$col] += 1;
            } elsif ( $action eq 'turn off' ) {
                $grid->[$row][$col] = ($grid->[$row][$col] > 0) ? $grid->[$row][$col] - 1 : 0;
            } elsif ( $action eq 'toggle' ) {
                $grid->[$row][$col] += 2;
           }
        }
    }
}

sub sum {
    my ($grid) = @_;
    my $sum = 0;

    for my $row (0..999) {
        for my $col (0..999) {
            $sum += $grid->[$row][$col];
        }
    }
    return $sum;
}

sub process {
    my ($filename) = @_;
    my $grid = init_grid();

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        if ( $line =~ /(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/ ) {
            my ($command, $r1, $c1, $r2, $c2) = ($1, $2, $3, $4, $5);
            perform($grid, $command, $r1, $c1, $r2, $c2);
        } else {
            next;
        }
    }
    close($infile);

    return sum($grid);
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();

File: python/2015/day06b/day06b.py
#!/usr/bin/env python3

from collections import defaultdict
import re
import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def perform(grid, action, r1, c1, r2, c2):
    for row in range(r1, r2+1):
        for col in range(c1, c2+1):
            if action == 'turn on':
                grid[(row, col)] += 1
            elif action == 'turn off':
                grid[(row, col)] -= 1 if grid[(row, col)] > 0 else grid[(row, col)]
            elif action == 'toggle':
                grid[(row, col)] += 2

def process(filename):
    grid = defaultdict(int)
    pattern = re.compile(r'(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)')
    with open(filename, 'r') as infile:
        for line in infile:
            matches = pattern.match(line)
            if matches != None:
                action = matches[1]
                r1 = int(matches[2])
                c1 = int(matches[3])
                r2 = int(matches[4])
                c2 = int(matches[5])
                perform(grid, action, r1, c1, r2, c2)
    return sum(grid.values())

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()
